src/ModelObjects/component.rs

```diff
@@ -255,6 +255,236 @@ impl Component {
         self.output_edges = Some(o_edges);
         self.input_edges = Some(i_edges);
     }
+
+    /// Function for reducing the clocks found on the component.
+    /// Unused clocks and "duplicate" clocks (clocks that are never reset)
+    /// and then remove them.
+    pub fn reduce_clocks(&mut self, redundant_clocks: &Vec<RedundantClock>) {
+        for clock in redundant_clocks {
+            match &clock.reason {
+                ClockReductionReason::Duplicate(global) => self.replace_clock(clock, global),
+                ClockReductionReason::Unused => self.remove_clock(&clock.updates),
+            }
+
+            let clock_val = *self
+                .declarations
+                .clocks
+                .get(clock.clock.as_str())
+                .unwrap_or_else(|| panic!("Clock {} is not in the declarations", clock.clock));
+            self.declarations
+                .clocks
+                .values_mut()
+                .filter(|val| **val > clock_val)
+                .for_each(|val| *val -= 1);
+            self.declarations.clocks.remove(clock.clock.as_str());
+        }
+    }
+
+    ///Used to find redundant clocks - checks for unused and duplicates clocks.
+    ///
+    /// Returns [`Vec<RedundantClock>`] with all found redundant clock.
+    /// If no redundant clocks found the vector will be empty
+    pub(crate) fn find_redundant_clocks(&self) -> Vec<RedundantClock> {
```

Comment:
The complexity of this function is very big. If it is possible it would be nice if you could factor out the body of this function into smaller functions

Reply:
Better now?

Reply:
yes ltgm